src/NLog/Layouts/Typed/Layout.cs

```diff
@@ -0,0 +1,350 @@
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Globalization;
+using System.Text;
+using JetBrains.Annotations;
+using NLog.Common;
+using NLog.Config;
+using NLog.Internal;
+
+namespace NLog.Layouts
+{
+    /// <summary>
+    /// Layout with a simple value (e.g. int) or a layout which results in a simple value (e.g. ${counter})
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class Layout<T> : Layout, IRenderable<T>
+    {
+        // ReSharper disable StaticMemberInGenericType - this is safe, static ctor is called for every generic type
+        [NotNull] private static readonly Type Type;
+        [NotNull] private static readonly string TypeNamed;
+        // ReSharper restore StaticMemberInGenericType
+
+        [CanBeNull] private readonly Layout _layout;
+        private readonly T _fixedValue;
+
+        /// <inheritdoc />
+        public Layout(T value)
+        {
+            _fixedValue = value;
+            IsFixed = true;
+            _layout = null;
+        }
+
+        /// <inheritdoc />
+        public Layout(Layout layout)
+        {
+            IsFixed = TryGetFixedValue(layout, out _fixedValue);
+            _layout = layout;
+        }
+
+        static Layout()
+        {
+            var type = typeof(T);
+            Type = Nullable.GetUnderlyingType(type) ?? type;
+            TypeNamed = type.Name;
+        }
+
+        /// <summary>
+        /// Is fixed value?
+        /// </summary>
+        public bool IsFixed { get; }
+
+        #region Overrides of TypedLayout<T>
+
+        private static string ValueToString(T value, CultureInfo cultureInfo)
+        {
+            if (value is IConvertible convertible)
+            {
+                return convertible.ToString(cultureInfo);
+            }
+
+            return value?.ToString();
+        }
+
+        private bool TryParse(string text, out T value, T defaultValue)
+        {
+            return TryConvertTo(text, out value, defaultValue);
+        }
+
+        private bool TryConvertTo(object raw, out T value, T defaultValue)
+        {
+            if (raw == null)
+            {
+                value = defaultValue;
+                return false;
+            }
+
+            var cultureInfo = CultureInfo.CurrentCulture;
+            try
+            {
+                var converter = ResolveService<IPropertyTypeConverter>() ?? PropertyTypeConverter.Instance;
+                var convertedValue = converter.Convert(raw, Type, null, cultureInfo);
+                if (convertedValue is T goodValue)
+                {
+                    value = goodValue;
```

Comment:
Should cache the raw-value and the good-value. And when trying to convert the same raw-value again, then just return the cached good-value (instead of converting one more time). Will reduce the overhead of converting the same comma-seperated-string into a List-of-items.

Reply:
To make it really nice. Then do like `FilePathLayout` where it skips StringBuilder.ToString() when it sees it contains the same raw-value as before.

Reply:
I like the cache idea, but I propose to do that in a follow-up PR. Im also convinced that's more your expertise :)

Reply:
Like when people complete their task, and does some benchmarks. Remember you want to use this everywhere, so you should make sure the idea works.

Reply:
If I add a benchmark, then to what should I compare to?

Reply:
Note, this isn't working yet "Will reduce the overhead of converting the same comma-seperated-string into a List-of-items.".  That's https://github.com/NLog/NLog/issues/3994

But agree, will look in the cache here

Reply:
Added a cache. it as simple on (last used), and new update will overwrite old values. Hope you like it. Maybe need some locks for updating and reading the cache